OPAL/tac/src/main/scala/org/opalj/tac/cg/android/AndroidICCAnalysis.scala

```diff
@@ -0,0 +1,1467 @@
+                    intent.categories += findString(tacCode.stmts(defSiteFirstParam).asAssignment).getOrElse(undef)
+                }
+                case `setData` ⇒ intent.iData = checkData(0, tacCode, parameter)
+                case `setDataAndType` ⇒
+                    intent.iData = checkData(0, tacCode, parameter)
+                    val defSite = parameter.last.asVar.definedBy.head
+                    if (defSite > -1) {
+                        intent.iType = findString(tacCode.stmts(defSite).asAssignment).getOrElse(undef)
+                    }
+                case `setDataAndNormalize` ⇒ intent.iData = checkData(0, tacCode, parameter)
+                case `setDataAndTypeAndNormalize` ⇒
+                    intent.iData = checkData(0, tacCode, parameter)
+                    val defSite = parameter.last.asVar.definedBy.head
+                    if (defSite > -1) {
+                        intent.iType = findString(tacCode.stmts(defSite).asAssignment).getOrElse(undef)
+                    }
+                case `setType` ⇒ if (defSiteFirstParam > -1) {
+                    intent.iType = findString(tacCode.stmts(defSiteFirstParam).asAssignment).getOrElse(undef)
+                }
+                case `setTypeAndNormalize` ⇒ if (defSiteFirstParam > -1) {
+                    intent.iType = findString(tacCode.stmts(defSiteFirstParam).asAssignment).getOrElse(undef)
+                }
+                case `setPackage` ⇒ if (defSiteFirstParam > -1) {
+                    intent.iPackage = findString(tacCode.stmts(defSiteFirstParam).asAssignment).getOrElse(undef)
+                }
+            }
+        } else if (activityStartMethods.contains(name)) {
+            intent.pc = pc
+            intent.componentTypes += activity
+            intent.calledMethods += onCreate
+        } else if (serviceMethods.contains(name)) {
+            intent.pc = pc
+            intent.componentTypes += service
+            name match {
+                case `startService`  ⇒ intent.calledMethods ++= ListBuffer(onCreate, onStartCommand)
+                case `bindService`   ⇒ intent.calledMethods ++= ListBuffer(onCreate, onBind, onRebind)
+                case `stopService`   ⇒ intent.calledMethods += stopSelf
+                case `unbindService` ⇒ intent.calledMethods += onUnbind
+            }
+        } else if (broadcastReceiverStartMethods.contains(name)) {
+            intent.pc = pc
+            intent.componentTypes += receiver
+            intent.calledMethods ++= ListBuffer(onReceive, onBroadcastReceive)
+        }
+        None
+    }
+
+    /**
+     * Matches found intents with intent filters and writes the results to 'edges'.
+     * Uses the results from 'searchIntentsAndCallbacks'.
+     */
+    def intentMatching(): Unit = {
+        //ordering Filters in maps for matching
+        val filterMap: mutable.Map[String, mutable.Map[String, mutable.Map[ListBuffer[String], ListBuffer[IntentFilter]]]] = mutable.Map.empty
```

Comment:
Please avoid all of these mutable data structures. Also, consider making the method more functional, not accessing and writing global state.

Reply:
Is it better to use 'var filterMap: immutable.Map[...]'?

Reply:
Yes, it usually is. In many cases, you can even have an immutable data structure in a val if you restructure your code to build the data structure in a single pass. Might not easily be possible here, though.